Providers/Profile.swift

```diff
@@ -509,43 +513,101 @@ open class BrowserProfile: Profile {
 
-            return self.remoteClientsAndTabs.getClients().bind { result in
-                if let error = result.failureValue { return deferMaybe(error) }
+    func getRustTabsWithClients() -> Deferred<Maybe<[ClientAndTabs]>> {
+        logger.log("Getting all tabs and clients", level: .debug, category: .tabs)
 
-                guard let clients = result.successValue else { return deferMaybe([]) }
+        guard let accountManager = self.rustFxA.accountManager.peek(),
+              let state = accountManager.deviceConstellation()?.state() else {
+            return deferMaybe([])
+        }
 
-                let clientAndTabs: [ClientAndTabs] = clientRemoteTabs.map { record in
-                    // We check if the application services clientId matches any client
-                    // GUID. If a client is found we return a record, otherwise we
-                    // continue to the next application services record.
-                    let localClient = clients.first(where: { $0.guid == record.clientId })
+        let remoteDeviceIds: [String] = state.remoteDevices.compactMap {
+            guard $0.capabilities.contains(.sendTab) else { return nil }
+            return $0.id
+        }
 
-                    if let client = localClient {
-                        return record.toClientAndTabs(client: client)
-                    }
+        let clientAndTabs = tabs.getRemoteClients(remoteDeviceIds: remoteDeviceIds)
+        return clientAndTabs
+    }
 
-                    self.logger.log("Could not find client data for appservices client ID \(record.clientId).",
-                                    level: .debug,
-                                    category: .tabs)
-                    return nil
-                }.compactMap { $0 }
+    func getTabsWithNativeClients() -> Deferred<Maybe<[ClientAndTabs]>> {
+        // Retrieving tabs with clients is handled differently depending on whether the
+        // rust sync manager component is being used. The `BrowserSyncManager` class syncs
+        // tabs with the rust tabs component and clients with the `ClientSynchronizer`
+        // class. So in order to ensure accurate tabs with clients data is returned while
+        // using these classes for syncing, tab records are pulled from the component's
+        // database table and filtered by the client records in the iOS BrowserDB database
+        // table.
+
+        // The new `RustSyncManager` class has no reliance on the `ClientSynchronizer` or
+        // the clients table. So when it's being used tabs with clients can be pulled from
+        // the tab component's database table and filtered by the remote devices in the
+        // user's device constellation. Once the sync manager rollout is complete this
+        // will be the way this data is retrieved unconditionally.
+        if rustSyncManagerEnabled {
+            return getRustTabsWithClients()
+        } else {
+            return tabs.getAll().bind { tabsResult in
```

Comment:
This branch of the `if` statement is pre-existing (see [here](https://github.com/mozilla-mobile/firefox-ios/blob/5a0ed3c1967400758241e1d410853d96199aad6d/Providers/Profile.swift#L520)).